src/graphmol/rw_mol.rs

```diff
@@ -48,11 +48,17 @@ impl RWMol {
         ROMol { ptr }
     }
 
-    pub fn from_smarts(smarts: &str) -> Result<Self, Box<dyn std::error::Error>> {
+    pub fn from_smarts(smarts: &str) -> Result<Self, &str> {
+        if smarts.is_empty(){ return Err("empty SMARTS"); }
```

Comment:
As I do not study chemistry myself, but informatics, I am not sure, if this is the wanted behavior or if you'd prefer a wildcard here or if maybe an empty string is fine.

Reply:
I think we probably don't want to allow an empty smarts, so this works for me!

Reply:
Thank you so much for contributing this PR! It's close but I think a little guidance would help. First off, it's good to reiterate the structure of this git repository. It contains two rust crates: rdkit and rdkit-sys. The sys crate is where pointers are built and where rust meets the C++ code of rdkit. You want all exception handling to happen in the sys crate. The `rdkit` crate is for high-level convenience. Every pointer type from `rdkit-sys` is wrapped in a struct -- with hidden pointers, as you've seen, we wan't to create a "nanny state" where our users can't do unsafe things. I believe you want to take a look at the sys crate as part of your bug fix.

Now for my feedback on this code snippet:

Using the `Result` type is absolutely the right way to go -- we need to discriminate between good values and error values. My only objection is the type on the error path, `&str` is not very Rust-y. We want to avoid "stringly-typed" data and instead use _more_ enums.

See https://github.com/rdkit-rs/rdkit/blob/f6da32a5838b980f60baaaac650fa2dc38b8fb75/src/graphmol/ro_mol.rs#L21-L34 for an example of how we build enum errors when working with SMILES. I believe you want to make a `RWMolError` enum and instantiate it in your error path.

My next concern is performing condition checking in `rw_mol.rs` itself -- in general I want rdkit (c++) to drive what happens. It's very possible in the future our friends at the RDKit project could change how empty strings behave (although unlikely) and I don't want the Rust logic to be at odds with whatever they decide. Instead, check if the returned pointer is null and if so, build the appropriate enum.

Reply:
Thank you so much for the detailed feedback! You are totally right about the enums and I am going to adapt it to the handling of errors in ro_mol.rs. 
Concerning the empty SMARTS, I checked what happens in Python and they return a valid Mol object that doesn't have any contents (so not a nullpointer, but empty). So maybe we should go for that, too?

Reply:
I prefer to model what the C++ library does -- it does mean that python folks are going to have to rethink how they use rdkit. The rdkit-rs implementation puts errors upfront and in your face. Clear error propagation is important for our users of the [the cheminee search engine](https://github.com/assaydepot/cheminee), for example. We don't want too much "magic".

So think of this project as C++ flavored RDKit 😄. Hopefully you are familiar with [the autogenerated rdkit docs](https://www.rdkit.org/docs/cppapi/index.html).

Reply:
I had a deeper look at the rdkit-Documentation, though I must say I have never dealt with C++. I still found an interesting comment [in the actual C++ code](https://github.com/rdkit/rdkit/blob/e605bf04ca8fcc67aa94f3843c6fe6697ed70681/Code/GraphMol/SmilesParse/SmilesParse.cpp#L224C3-L224C44) stating: `// empty strings produce empty molecules`. So maybe having an error in Rust at this part would not really fit?

Reply:
I agree, please address my last comment and I'm happy to merge this and push it up to crates.io. Thanks for your perseverance and patience!